proposals/NNNN-opacity-micromaps.md

```diff
@@ -0,0 +1,384 @@
+See the Opacity micromaps section in the [Raytracing spec][dxr-omm]
+for more details on the overall feature and D3D runtime context.
+
+## Motivation
+
+There needs to be a way to enable OMM through subobjects in HLSL, equivalent
+to [`D3D12_RAYTRACING_PIPELINE_FLAG_ALLOW_OPACITY_MICROMAPS`][dxr-flags] in the
+API.
+
+Additionally, the runtime allows instances to be tagged with
+`D3D12_RAYTRACING_INSTANCE_FLAG_FORCE_OMM_2_STATE`, but allowing this to be
+specified at ray tracing time in a shader would provide flexibility for
+developers.
+
+## Proposed solution
+
+A new `RAYTRACING_PIPELINE_FLAG` is proposed to enable OMM in HLSL:
+`RAYTRACING_PIPELINE_FLAG_ALLOW_OPACITY_MICROMAPS`.
+These flags are used in the [Raytracing pipeline config1][pipeline-config]
+subobject in HLSL.
+
+A new `RAY_FLAG` is proposed that allows ray tracing in shaders to force OMM
+2-state for a particular trace: `RAY_FLAG_FORCE_OMM_2_STATE`.
+
+## Detailed design
+
+### HLSL Additions
+
+Add a built-in `RAYTRACING_PIPELINE_FLAG_ALLOW_OPACITY_MICROMAPS` to the
+`RAYTRACING_PIPELINE_FLAG` flags in HLSL to enable OMM.
+Add built-in `RAY_FLAG_FORCE_OMM_2_STATE` to `RAY_FLAG` flags in HLSL.
+
+`RAYTRACING_PIPELINE_FLAG` flags are used in the
+[Raytracing pipeline config1][pipeline-config]
+subobject in HLSL.  When `RAYTRACING_PIPELINE_FLAG_ALLOW_OPACITY_MICROMAPS` is
+used in the current pipeline configuration, the pipeline supports Opacity
+Micromaps. If a triangle with an OMM is encountered during traversal with this
+flag cleared, behavior is undefined. This flag should not be set if there are
+no OMMs present, since it may incur a small penalty on traversal performance
+overall.  See the D3D12 flag definition for more details:
+[`D3D12_RAYTRACING_PIPELINE_FLAG_ALLOW_OPACITY_MICROMAPS`][dxr-flags].
+
+Add a built-in `RAY_FLAG_FORCE_OMM_2_STATE` flag to the `RAY_FLAG` flags.
+`RAY_FLAG` flags are only meaningful when used in the `RayFlags` parameter of
+the [TraceRay()][trace-ray] or [RayQuery::TraceRayInline()][rq-trace]
+intrinsic functions, or in the `RAY_FLAGS` template argument of a
+[RayQuery][ray-query] object.  When used with [Opacity Micromaps][dxr-omm]
+(`RAYTRACING_PIPELINE_FLAG_ALLOW_OPACITY_MICROMAPS` or
+[`D3D12_RAYTRACING_PIPELINE_FLAG_ALLOW_OPACITY_MICROMAPS`][dxr-flags]),
+the raytracing pipeline will ignore the Unknown state and only consider the
+Transparent/Opaque bit for all 4-state Opacity Micromaps encountered during
+traversal. This flag has no effect if
+`D3D12_RAYTRACING_INSTANCE_FLAG_DISABLE_OMMS` is set on the instance, or if
+Opacity Micromaps are globally not set, not allowed, or not supported.
+If set, this flag will be present in the returned flags of the
+[RayFlags()][rayflags] intrinsic function.
+
+In HLSL under DXC, these are defined as static const uint values:
+
+```hlsl
+static const uint RAYTRACING_PIPELINE_FLAG_ALLOW_OPACITY_MICROMAPS = 0x400;
+static const uint RAY_FLAG_FORCE_OMM_2_STATE = 0x400;
+```
```

Comment:
The equivalent of PIPELINE_FLAG_ALLOW_OPACITY_MICROMAPS is needed for inline raytracing as well (e.g. RayQuery), which isn't affected by RAYTRACING_PIPELINE_FLAG*.

A proposed solution is a new template parameter to the RayQuery object.  Pasting the text already proposed for the full DXR spec for this:

### RayQuery flags

RayQuery flags are optionally specified as a second template parameter to [RayQuery](#rayquery) objects to enable static specialization - runtime behavior of [RayQuery::TraceRayInline()](#rayquery-tracerayinline) takes on the OR of dynamic and static/template ray flags as well as rayquery-specific flags.

```C++
enum RAYQUERY_FLAG : uint
{
    RAYQUERY_FLAG_NONE = 0x00, // default
    RAYQUERY_FLAG_ALLOW_OPACITY_MICROMAPS = 0x01,
};
```

Value                               | Definition
---------                           | ----------
`RAYQUERY_FLAG_NONE` | No options selected.
`RAYQUERY_FLAG_ALLOW_OPACITY_MICROMAPS` | If this flag is specified, the rayquery object supports Opacity Micromaps. If a triangle with an OMM is encountered during traversal with this flag cleared, behavior is undefined. This flag should not be set if there are no OMMs present, since it may incur a small penalty on traversal performance overall.  This is the rayquery equivalent for [TraceRayInline()](#rayquery-tracerayinline) of what the [D3D12_RAYTRACING_PIPELINE_FLAG_ALLOW_OPACITY_MICROMAPS](#d3d12_raytracing_pipeline_flags) flag accomplishes for non-inline [TraceRay()](#traceray).
`D3D12_RAYTRACING_PIPELINE_FLAG_ALLOW_OPACITY_MICROMAPS` | If this flag is specified, the pipeline supports Opacity Micromaps. This flag is part of the [Opacity Micromaps](#opacity-micromaps) feature.

Reply:
Is ok to just get the proposal in and then do a follow-up PR to make changes to it please?